File: app/profile_parser.py
from datetime import datetime, timezone
from typing import Any

# Bump when parsing logic changes, so cached records stay traceable to the parser that produced
# them and re-parsed results are distinguishable from live ones.
PARSER_VERSION = 1


def _entities(response: dict[str, Any], type_suffix: str) -> list[dict[str, Any]]:
    """Voyager responses are a flat `included` list of typed entities rather than one nested
    profile object; each entity carries a `$type` like
    com.linkedin.voyager.identity.profile.Position. Filtering by type suffix (rather than an
    exact string) keeps this working if LinkedIn re-namespaces the entities."""
    return [e for e in response.get("included", []) if str(e.get("$type", "")).endswith(type_suffix)]


def _first(response: dict[str, Any], type_suffix: str) -> dict[str, Any] | None:
    found = _entities(response, type_suffix)
    return found[0] if found else None


def _text(value: Any) -> str | None:
    return value if isinstance(value, str) and value else None


def _date(value: Any) -> dict[str, int | None] | None:
    if not isinstance(value, dict):
        return None
    if "month" not in value and "year" not in value:
        return None
    return {"month": value.get("month"), "year": value.get("year")}


def _time_period(entity: dict[str, Any]) -> tuple[Any, Any]:
    period = entity.get("timePeriod") or {}
    return _date(period.get("startDate")), _date(period.get("endDate"))


def _profile_images(profile: dict[str, Any] | None) -> list[dict[str, Any]]:
    if not profile:
        return []
    vector = (
        (profile.get("profilePicture") or {})
        .get("displayImageReference", {})
        .get("vectorImage")
    )
    if not vector or not vector.get("rootUrl"):
        return []

    images = []
    for artifact in vector.get("artifacts") or []:
        segment = artifact.get("fileIdentifyingUrlPathSegment")
        if not segment:
            continue
        images.append(
            {
                "url": f"{vector['rootUrl']}{segment}",
                "width": artifact.get("width"),
                "height": artifact.get("height"),
            }
        )
    return images


def parse_profile_view(raw: Any, public_identifier: str, profile_url: str) -> dict[str, Any]:
    response: dict[str, Any] = raw if isinstance(raw, dict) else {}
    profile = _first(response, "identity.profile.Profile") or {}

    first_name = _text(profile.get("firstName"))
    last_name = _text(profile.get("lastName"))
    full_name = " ".join(p for p in (first_name, last_name) if p) or None

    experience = []
    for position in _entities(response, "Position"):
        start, end = _time_period(position)
        experience.append(
            {
                "title": _text(position.get("title")),
                "companyName": _text(position.get("companyName")),
                "location": _text(position.get("locationName")),
                "description": _text(position.get("description")),
                "startDate": start,
                "endDate": end,
                "isCurrent": end is None and start is not None,
            }
        )

    education = []
    for school in _entities(response, "Education"):
        start, end = _time_period(school)
        education.append(
            {
                "schoolName": _text(school.get("schoolName")),
                "degreeName": _text(school.get("degreeName")),
                "fieldOfStudy": _text(school.get("fieldOfStudy")),
                "startYear": (start or {}).get("year"),
                "endYear": (end or {}).get("year"),
                "description": _text(school.get("description")),
            }
        )

    certifications = []
    for cert in _entities(response, "Certification"):
        start, end = _time_period(cert)
        certifications.append(
            {
                "name": _text(cert.get("name")),
                "authority": _text(cert.get("authority")),
                "startDate": start,
                "endDate": end,
            }
        )

    return {
        "publicIdentifier": public_identifier,
        "profileUrl": profile_url,
        "firstName": first_name,
        "lastName": last_name,
        "fullName": full_name,
        "headline": _text(profile.get("headline")),
        "location": _text(profile.get("geoLocationName")) or _text(profile.get("locationName")),
        "about": _text(profile.get("summary")),
        "profileImages": _profile_images(profile),
        "experience": experience,
        "education": education,
        "skills": [
            {"name": _text(s.get("name")), "endorsementCount": s.get("endorsementCount")}
            for s in _entities(response, "Skill")
            if _text(s.get("name"))
        ],
        "certifications": certifications,
        "languages": [
            {"name": _text(l.get("name")), "proficiency": _text(l.get("proficiency"))}
            for l in _entities(response, "Language")
            if _text(l.get("name"))
        ],
        "fetchedAt": datetime.now(timezone.utc).isoformat(),
    }

File: src/profileParser.ts
import type {
  CertificationEntry,
  EducationEntry,
  ExperienceEntry,
  LanguageEntry,
  LinkedInProfile,
  ProfileImage,
  SkillEntry,
} from "./types";

// Bump whenever parsing logic changes so cached/stored records can be traced back to the parser
// version that produced them, and so reparse() results are distinguishable from live ones.
export const PARSER_VERSION = 1;

// LinkedIn's voyager responses are a flat "included" array of typed entities (Profile, Position,
// Education, Skill, ...) rather than one nested object. This shape is undocumented and has shifted
// across LinkedIn releases before — see README "Known limitations".
interface VoyagerEntity {
  $type?: string;
  entityUrn?: string;
  [key: string]: unknown;
}

interface VoyagerResponse {
  included?: VoyagerEntity[];
}

function entitiesOfType(response: VoyagerResponse, typeSuffix: string): VoyagerEntity[] {
  return (response.included ?? []).filter((entity) => entity.$type?.endsWith(typeSuffix));
}

function firstEntityOfType(response: VoyagerResponse, typeSuffix: string): VoyagerEntity | undefined {
  return entitiesOfType(response, typeSuffix)[0];
}

function asString(value: unknown): string | null {
  return typeof value === "string" && value.length > 0 ? value : null;
}

function asNumber(value: unknown): number | null {
  return typeof value === "number" ? value : null;
}

function parseDate(value: unknown): { month: number | null; year: number | null } | null {
  if (!value || typeof value !== "object") return null;
  const date = value as { month?: number; year?: number };
  if (date.month === undefined && date.year === undefined) return null;
  return { month: date.month ?? null, year: date.year ?? null };
}

function parseTimePeriod(entity: VoyagerEntity): {
  startDate: { month: number | null; year: number | null } | null;
  endDate: { month: number | null; year: number | null } | null;
} {
  const timePeriod = entity.timePeriod as { startDate?: unknown; endDate?: unknown } | undefined;
  return {
    startDate: parseDate(timePeriod?.startDate),
    endDate: parseDate(timePeriod?.endDate),
  };
}

function parseProfileImages(profile: VoyagerEntity | undefined): ProfileImage[] {
  const displayImageReference = (profile as { profilePicture?: { displayImageReference?: unknown } } | undefined)
    ?.profilePicture?.displayImageReference as
    | { vectorImage?: { rootUrl?: string; artifacts?: Array<{ width?: number; height?: number; fileIdentifyingUrlPathSegment?: string }> } }
    | undefined;

  const vectorImage = displayImageReference?.vectorImage;
  if (!vectorImage?.rootUrl || !vectorImage.artifacts) return [];

  return vectorImage.artifacts
    .filter((artifact) => artifact.fileIdentifyingUrlPathSegment)
    .map((artifact) => ({
      url: `${vectorImage.rootUrl}${artifact.fileIdentifyingUrlPathSegment}`,
      width: artifact.width,
      height: artifact.height,
    }));
}

function parseExperience(response: VoyagerResponse): ExperienceEntry[] {
  return entitiesOfType(response, "Position").map((position) => {
    const { startDate, endDate } = parseTimePeriod(position);
    return {
      title: asString(position.title),
      companyName: asString(position.companyName),
      location: asString(position.locationName),
      description: asString(position.description),
      startDate,
      endDate,
      isCurrent: endDate === null && startDate !== null,
    };
  });
}

function parseEducation(response: VoyagerResponse): EducationEntry[] {
  return entitiesOfType(response, "Education").map((education) => {
    const { startDate, endDate } = parseTimePeriod(education);
    return {
      schoolName: asString(education.schoolName),
      degreeName: asString(education.degreeName),
      fieldOfStudy: asString(education.fieldOfStudy),
      startYear: startDate?.year ?? null,
      endYear: endDate?.year ?? null,
      description: asString(education.description),
    };
  });
}

function parseSkills(response: VoyagerResponse): SkillEntry[] {
  return entitiesOfType(response, "Skill")
    .map((skill) => ({
      name: asString(skill.name) ?? "",
      endorsementCount: asNumber(skill.endorsementCount),
    }))
    .filter((skill) => skill.name.length > 0);
}

function parseCertifications(response: VoyagerResponse): CertificationEntry[] {
  return entitiesOfType(response, "Certification").map((cert) => {
    const { startDate, endDate } = parseTimePeriod(cert);
    return {
      name: asString(cert.name) ?? "",
      authority: asString(cert.authority),
      startDate,
      endDate,
    };
  });
}

function parseLanguages(response: VoyagerResponse): LanguageEntry[] {
  return entitiesOfType(response, "Language").map((language) => ({
    name: asString(language.name) ?? "",
    proficiency: asString(language.proficiency),
  }));
}

export function parseProfileView(raw: unknown, publicIdentifier: string, profileUrl: string): LinkedInProfile {
  const response = raw as VoyagerResponse;
  const profile = firstEntityOfType(response, "identity.profile.Profile");

  const firstName = asString(profile?.firstName);
  const lastName = asString(profile?.lastName);

  return {
    publicIdentifier,
    profileUrl,
    firstName,
    lastName,
    fullName: firstName || lastName ? [firstName, lastName].filter(Boolean).join(" ") : null,
    headline: asString(profile?.headline),
    location: asString(profile?.geoLocationName) ?? asString(profile?.locationName),
    about: asString(profile?.summary),
    profileImages: parseProfileImages(profile),
    experience: parseExperience(response),
    education: parseEducation(response),
    skills: parseSkills(response),
    certifications: parseCertifications(response),
    languages: parseLanguages(response),
    fetchedAt: new Date().toISOString(),
  };
}
